photon-ml/src/main/scala/com/linkedin/photon/ml/data/KeyValueScore.scala

```diff
@@ -63,14 +63,21 @@ protected[ml] class KeyValueScore(val scores: RDD[(Long, ScoredGameDatum)]) exte
   def binaryOp(
     op: (ScoredGameDatum, ScoredGameDatum) => ScoredGameDatum,
     that: KeyValueScore): KeyValueScore = {
-    val scores =
-      this.scores
-        .fullOuterJoin(that.scores)
-        .mapValues {
-          case (Some(thisScore), Some(thatScore)) => op(thisScore, thatScore)
-          case (Some(thisScore), None) => thisScore
-          case (None, Some(thatScore)) => thatScore
+    val scores = this.scores
+      .cogroup(that.scores)
+      .flatMapValues { case (s1It, s2It) =>
+        assert(
+          s1It.size <= 1 && s2It.size <= 1,
+          s"More than one scored data found in co-grouped RDD for an id. Found ${s1It.size} records on the left, " +
+            s"${s2It.size} on the right")
+        if (s1It.isEmpty) {
+          Seq(op(s2It.head.getZeroScoreDatum, s2It.head))
+        } else if (s2It.isEmpty) {
+          Seq(op(s1It.head, s1It.head.getZeroScoreDatum))
+        } else {
+          Seq(op(s1It.head, s2It.head))
         }
+      }
```

Comment:
Does this approach perform significantly better? I find this result a little surprising, and it might need to inform some changes down the road on how we organize chained operations. Can we publish the benchmarking scripts and analysis on a wiki or something? @fastier-li